### ISSUE TITLE
Redraw errors when using retina screens

### ISSUE THREAD

**Issue by millifoo** [plexinc/plex-media-player#10]
Howdy - just installed the preview (file says 1.0.0.5-53192cb0,  when it starts it says 2.4.27).

Video of what I'm seeing: https://dl.dropboxusercontent.com/u/46967890/plex-glitch.mov

(I'll leave this there for a couple weeks, but please try and copy it -- I tried attaching it, it's too big :-).

The first glitch can be seen early on - mostly just flashing when things are repainting. The second is near the end when I try to exist - should be obvious when you view it.

Hardware: 
  Model Name:	2013 MacBook Pro
  Model Identifier:	MacBookPro10,1
  Processor Name:	Intel Core i7
  Processor Speed:	2.7 GHz
  Number of Processors:	1
  Total Number of Cores:	4
  L2 Cache (per Core):	256 KB
  L3 Cache:	6 MB
  Memory:	16 GB
  Boot ROM Version:	MBP101.00EE.B09
  HD:  500 GB SSD

NVIDIA GeForce GT 650M:
  Chipset Model:	NVIDIA GeForce GT 650M
  Type:	GPU
  Bus:	PCIe
  PCIe Lane Width:	x8
  VRAM (Total):	1024 MB
  Vendor:	NVIDIA (0x10de)
  Device ID:	0x0fd5
  Revision ID:	0x00a2
  ROM Revision:	3688
  gMux Version:	3.2.19 [3.2.8]

email: millifoo at gmail dot com

**Comment by tru** [comment]
Is that a retina macbook pro? What's your desktop scaling set to? We have seen issues with retina macs when the setting is "more space". It's something we are looking into.

**Comment by donmc123** [comment]
Same here on Mac Pro with El Capitan with 4K display. Pretty much at all resolutions I've tried

**Comment by tru** [comment]
Yeah it seems to be related to screens with a LOT of pixels. We will try to solve this asap.

**Comment by tru** [comment]
Most likely this is because something in the webenginecore. I will try to see if there are any open issues in Qt for this.

**Comment by tru** [comment]
Putting it in 1.0.1 milestone for now, but I am sceptical we will get it done unless there is something obvious we can pull from upstream Qt.

**Comment by tru** [comment]
Jocelyn suggested to test with `--disable-gpu` to see if that fixed the issue, it did not.

**Comment by millifoo** [comment]
> Is that a retina macbook pro?

Yes - correct, retina macbook with "more space" chosen. Changing over to "default" eliminates the problem.

**Comment by tru** [comment]
Yeah we are looking into it. Might take a while since it seems to be a very annoying bug.

**Comment by sabiwa** [comment]
I can confirm this also happens on an retina 27 '' iMac.

**Comment by jturcotte** [comment]
I could reproduce the issue by forcing a higher devicePixelRatio on my mbp (the viewport scale would flicker, but I could still see the issue at least).

After investigation it seems like --disable-gpu is always forced once Chromium fails creating a GL context 4 times. I get an "invalid share context" error from CGLCreateContext in gfx::GLContextCGL::Initialize which I don't get with qmlscene, so the version + profile in your QWindow's pixel format is probably not compatible with the one Chromium tries to use. See GetPixelFormat in qtwebengine/src/3rdparty/chromium/ui/gl/gl_context_cgl.cc to see what they try creating.

Removing the QSurfaceFormat::setDefaultFormat change in pmp works around that issue, but my guess is that it won't be trivial to get Chromium to properly support GL 3.

After that it seems like this triggers a software limit on the resource count (textures associated with each tile) in CompositorOutputSurface::BindToClient. This was changed by <https://src.chromium.org/viewvc/chrome?revision=283370&view=revision>. Reverting this change and setting the memory limit to 256 * 1024 * 1024 seems to work around the issue for me, though this might trigger facing https://code.google.com/p/chromium/issues/detail?id=384654 at some point, so not perfect either.

**Comment by tru** [comment]
So to recap:

* Remove the QSurfaceFormat stuff from main that we force to use OpenGL 3.x and higher.
* Patch chromium to use a texture buffer size.

If the second problem is something that is truly a problem in chromium only, why do we only see this happening in PMP with WebEngine instead of in desktop chrome. Do they tweak these values?

**Comment by wm4** [comment]
Would disabling process separation work around the resource limit problem?

**Comment by tru** [comment]
Oh that's a good question as well. We have been thinking about disabling process separation anyway since it doesn't buy us anything and it potentially misses crash reports etc. But we where worried that this code path is not as well tested. Do you have any experience with this @jturcotte ?

**Comment by tru** [comment]
Confirmed that switching back to legacy GL context and adding `--single-process` fixes the redraw problems. BUT, instead I get a crash on exit:

```
Process:               Plex Media Player [52509]
Path:                  /Users/USER/Library/Caches/*/Plex Media Player.app/Contents/MacOS/Plex Media Player
Identifier:            tv.plex.player
Version:               1.0.2.dev-17f51ba4 (1.0.2.dev-17f51ba4)
Code Type:             X86-64 (Native)
Parent Process:        clion [30134]
Responsible:           Plex Media Player [52509]
User ID:               501

Date/Time:             2015-11-03 12:59:23.182 +0100
OS Version:            Mac OS X 10.11 (15A279b)
Report Version:        11
Anonymous UUID:        57306247-C06A-B3FB-5CF8-B02F1D5620CB

Sleep/Wake UUID:       160CAA6D-8E5F-4E11-982D-48EDEC467F6B

Time Awake Since Boot: 100000 seconds
Time Since Wake:       220 seconds

System Integrity Protection: enabled

Crashed Thread:        0  CrBrowserMain  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000029
Exception Note:        EXC_CORPSE_NOTIFY

VM Regions Near 0x29:
--> 
    __TEXT                 000000010529c000-0000000105615000 [ 3556K] r-x/rwx SM=COW  /Users/USER/Library/Caches/*/Plex Media Player.app/Contents/MacOS/Plex Media Player

Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   org.qt-project.Qt.QtWebEngineCore	0x0000000109710580 base::SupportsUserData::GetUserData(void const*) const + 32
1   org.qt-project.Qt.QtWebEngineCore	0x000000010a4bbd75 content::RenderProcessHostImpl::UnregisterHost(int) + 597
2   org.qt-project.Qt.QtWebEngineCore	0x000000010a4bb676 content::RenderProcessHostImpl::~RenderProcessHostImpl() + 422
3   org.qt-project.Qt.QtWebEngineCore	0x000000010a4bbe7e content::RenderProcessHostImpl::~RenderProcessHostImpl() + 14
4   org.qt-project.Qt.QtWebEngineCore	0x000000010a4bb34c content::RenderProcessHostImpl::ShutDownInProcessRenderer() + 732
5   org.qt-project.Qt.QtWebEngineCore	0x000000010a31374a content::BrowserMainLoop::ShutdownThreadsAndCleanUp() + 218
6   org.qt-project.Qt.QtWebEngineCore	0x000000010a315cf5 content::BrowserMainRunnerImpl::Shutdown() + 357
7   org.qt-project.Qt.QtWebEngineCore	0x000000010a3158f7 content::BrowserMainRunnerImpl::~BrowserMainRunnerImpl() + 39
8   org.qt-project.Qt.QtWebEngineCore	0x0000000108e167a0 WebEngineContext::~WebEngineContext() + 192
9   org.qt-project.Qt.QtWebEngineCore	0x0000000108e168b0 (anonymous namespace)::destroyContext() + 48
10  org.qt-project.QtCore         	0x0000000106ffc0e5 QCoreApplication::~QCoreApplication() + 69
11  org.qt-project.QtGui          	0x00000001068436e3 QGuiApplication::~QGuiApplication() + 499
12  tv.plex.player                	0x00000001052ab3df main + 5231 (main.cpp:271)
13  libdyld.dylib                 	0x00007fff9174c5ad start + 1
```

Most likely related to the single process invocation?

**Comment by jturcotte** [comment]
You only need to do one of the two, either use a GL 2 profile, either increase the resource count limit by patching Chromium. So in your case above, using a legacy GL context should also be enough to prevent seeing the issue.

I might be wrong but I think that the resource count, and also the fact that texture data is transferred using shared memory is not going to change even if you use --single-process. The switch uses threads instead of separate processes for components but still uses the same IPC mechanisms to communicate between them, since this is used by Google only as a testing function.

**Comment by tru** [comment]
Oh in that case I think it's easy to just drop down to legacy OpenGL contexts on Mac. Let me try just that.

**Comment by wm4** [comment]
We don't want to use legacy GL, as it complicates interaction between hw decoding and our GL video renderer. (IOSurfaces can be mapped as rectangle textures only, and rectangle textures behave differently in legavy GL.)

**Comment by tru** [comment]
better than redraw issues.

**Comment by wm4** [comment]
This will break hw decoding on OSX. Maybe this is acceptable, because there were other problems in certain cases?

**Comment by tru** [comment]
I made some tests:

With OpenGL v3 HW decoder is enabled and we use around 6-7% of CPU on my machine. The redraw errors makes the interface pretty unusable.

With OpenGL v2 the HW decoder is disabled and we are using around 45% CPU on my machine. The redraw errors are not there at all.

So yeah, ideally we want to get it working with OpenGL3.

As a fallback I think we can try to detect machines with this problem (check resolution or something?) and disable OpenGLv3 there, most of those machines should be fast enough for real-time playback without the hw decoder. But I consider this a suboptimal solution.

**Comment by tru** [comment]
Just for posterity, we are currently trying a patch to increase the relevant buffer in chromium so that we can continue to use modern OpenGL and hardware decoding.

**Comment by tru** [comment]
Rebuilding Qt on the CI and then kicking off a test build for you guys. But it seems to behave on my systems.

**Comment by tru** [comment]
Please try this build: https://nightlies.plex.tv/directdl/plexmediaplayer/dist-ninja/1.0.2.53-feaa648a/PlexMediaPlayer-1.0.2.53-feaa648a-macosx-x86_64.zip

It should fix the redraw errors.

**Comment by atrus05** [comment]
With that build I can no longer reproduce the issue on my Macbook Pro Retina with scaling set to highest. The posters still loads much slower compared to when I have it set to lower, but that I suspect is a much different issue :)

**Comment by chrishoage** [comment]
This is a little more laggy because I am recording, but the redraw issues persist even when I am not recording. 
https://www.dropbox.com/s/6ue478suciuzs5d/Redraw%20issues.mov?dl=0

log: https://www.dropbox.com/s/psgfdqqkq5oj519/Plex%20Media%20Player.log?dl=0

**Comment by tru** [comment]
@jturcotte what do you think. We run out of tiles again because of the even higher resolution?

**Comment by jturcotte** [comment]
Quite possible, but something looks overly small in this video. Tiles should be 256x256, the screen is 2160 px high according to the log and this should mean that we should be able to fit 8-9 tiles on the screen, but I can roughly count that ~15 tiles would fit in the screen's height. Is it possible that something is scaling the webview any more than it should on this setup?

I'm also not sure if the performance cost of rendering that many tiles on the screen is worth the extra sharpness. Would it be an option to cap the webview's resolution to 1080p and scale the result up to 2160 to avoid this issue and also improve the rendering speed? Or is speed usually not an issue at that resolution?

**Comment by tru** [comment]
We have had problems with the webview overflowing before. I wonder if that still happens @dogonwheels @mseeley

**Comment by mseeley** [comment]
> We have had problems with the webview overflowing before. I wonder if that still happens

No this shouldn't be an issue. The `body` has a fixed size and the app container is set to the same size and hides overflow. I'm not sure why the `body` doesn't also hide overflow; that's a little curious.

> Would it be an option to cap the webview's resolution to 1080p and scale the result up to 2160 to avoid this issue and also improve the rendering speed?

Would be super interesting to bake off the two approaches. I'm concerned that text would be blurry when scaled up 100%. Could be a decent stop gap until we can get perf and layer work addressed.

**Comment by tru** [comment]
I just enabled mouse scroll in PMP and can confirm that nothing seems to scroll like it did before. So it doesn't seem like we are overflowing the web view in anyway.

I say we go with a resolution clamp until we can find a better solution. That will be better than not working at all.

I think we can even have a higher clamp than 1080p, but make it configurable so we can test our way until our limit. I will discuss with Dom today.

**Comment by dogonwheels** [comment]
I've added a url param to clamp in https://github.com/plexinc/plex-web-client/tree/konvergo-clamp-resolution. Pass `?maxWidth=2160` or whatever.

While there, I made the scale based on the minimum of horizontal and vertical scale. I think we dodged this for now because we also had this logic in QML - but we shouldn't rely on that, I reckon.

I also added a minimum scale of 1 to prevent the flickering right now. Interestingly if you totally fix the value, it avoids the flicker of e.g. 0.5 to 1.0 to 0.5 - so we could potentially pass a fixed minimum, rather than check window size when < 1. Not sure.

One last note is that this obviously changes our `href` so influences anything we later build that relies on this. I suspect we'd bookmark/etc on state rather than full href, but still.
